pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,726 @@
+
+                let user_bid = auction
+                    .bidders
+                    .remove_bid(who.clone())
+                    .ok_or(Error::<T>::BidDoesNotExist)?;
+
+                T::Currency::transfer(&Self::account_id(), &who, user_bid.1, KeepAlive)?;
+                // emit bid removed event
+                Self::deposit_event(Event::BidRemoved {
+                    nft_id,
+                    marketplace_id: auction.marketplace_id,
+                    creator: who,
+                });
+                Ok(())
+            })?;
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn increase_bid(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                let auction = maybe_auction
+                    .as_mut()
+                    .ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+                // ensure the auction period has not ended
+                ensure!(auction.end_block > current_block, Error::<T>::AuctionEnded);
+
+                // ensure the bid is larger than the current highest bid
+                if let Some(current_highest_bid) = auction.bidders.get_highest_bid() {
+                    ensure!(amount > current_highest_bid.1, Error::<T>::InvalidBidAmount);
+                }
+
+                let user_bid = auction
+                    .bidders
+                    .remove_bid(who.clone())
+                    .ok_or(Error::<T>::BidDoesNotExist)?;
+
+                let amount_to_transfer = amount
+                    .checked_sub(&user_bid.1)
+                    .ok_or(Error::<T>::UnexpectedError)?;
+
+                T::Currency::transfer(&who, &Self::account_id(), amount_to_transfer, KeepAlive)?;
+
+                auction.bidders.insert_new_bid(who.clone(), amount);
+
+                // emit bid updated event
+                Self::deposit_event(Event::BidUpdated {
+                    nft_id,
+                    marketplace_id: auction.marketplace_id,
+                    creator: who,
+                });
+                Ok(())
+            })?;
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn buy_it_now(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // transfer funds from caller to pallet
+            T::Currency::transfer(&who, &Self::account_id(), amount, KeepAlive)?;
```

Comment:
This is unnecessary. We can transfer funds directly from the callers account instead of sending them first to the pallet's account.